qutebrowser/browser/qutescheme.py

```diff
@@ -158,6 +161,59 @@ def qute_bookmarks(_url):
     return 'text/html', html
 
 
+@add_handler('history')
+def qute_history(url):
+    """Handler for qute:history. Display history."""
+    # Get current date from query parameter, if not given choose today.
+    curr_date = datetime.date.today()
+    try:
+        query_date = QUrlQuery(url).queryItemValue("date")
+        if query_date:
+            curr_date = datetime.datetime.strptime(query_date, "%Y-%m-%d")
+            curr_date = curr_date.date()
+    except ValueError:
+        log.misc.debug("Invalid date passed to qute:history: " + query_date)
+
+    one_day = datetime.timedelta(days=1)
+    next_date = curr_date + one_day
+    prev_date = curr_date - one_day
+
+    def history_iter():
+        for item in objreg.get('web-history').history_dict.values():
```

Comment:
My CPU is spinning up quite a bit while browsing history, and I can see a performance improvement here: We don't need to do all this for the whole history! However, it gets a bit complicated, as we can only do this performance improvement in case we're on Python 3.5 or newer, where we can reverse an OrderedDict in-place. Otherwise, we'd need to copy the whole history into a list (the equivalent of `revered(list(history_iter()))` below), which would hurt performance again.

The idea is this: If we're on 3.5 or newer, we reverse the history in here, by doing something like:

```python
if sys.hexversion >= 0x03050000:
    history = reversed(objreg.get('web-history').history_dict.values())
    is_reversed = True
else:
    history = objreg.get('web-history').history_dict.values()
    is_reversed = True
```

That means we can now return from the generator if it is reversed and we found an entry which is too old, as they're only getting older.

However, you'd also need to adjust the code using the iterator to *not* do the reversing in this case. Not exactly straightforward, but I think it's worth the trouble as it's probably a big performance improvement.